resolve.go

```diff
@@ -44,141 +45,179 @@ func (r *MockBackend) LookupTXT(ctx context.Context, name string) ([]string, err
 	}
 }
 
-func Matches(maddr ma.Multiaddr) bool {
-	protos := maddr.Protocols()
-	if len(protos) == 0 {
-		return false
-	}
-
-	for _, p := range ResolvableProtocols {
-		if protos[0].Code == p.Code {
-			return true
+func Matches(maddr ma.Multiaddr) (matches bool) {
+	ma.ForEach(maddr, func(c ma.Component) bool {
+		switch c.Protocol().Code {
+		case DnsProtocol.Code, Dns4Protocol.Code, Dns6Protocol.Code, DnsaddrProtocol.Code:
+			matches = true
 		}
-	}
-
-	return false
+		return !matches
+	})
+	return matches
 }
 
 func Resolve(ctx context.Context, maddr ma.Multiaddr) ([]ma.Multiaddr, error) {
 	return DefaultResolver.Resolve(ctx, maddr)
 }
 
 func (r *Resolver) Resolve(ctx context.Context, maddr ma.Multiaddr) ([]ma.Multiaddr, error) {
-	if !Matches(maddr) {
-		return []ma.Multiaddr{maddr}, nil
-	}
-
-	protos := maddr.Protocols()
-	if protos[0].Code == Dns4Protocol.Code {
-		return r.resolveDns4(ctx, maddr)
-	}
-	if protos[0].Code == Dns6Protocol.Code {
-		return r.resolveDns6(ctx, maddr)
-	}
-	if protos[0].Code == DnsaddrProtocol.Code {
-		return r.resolveDnsaddr(ctx, maddr)
-	}
-
-	panic("unreachable")
-}
-
-func (r *Resolver) resolveDns4(ctx context.Context, maddr ma.Multiaddr) ([]ma.Multiaddr, error) {
-	value, err := maddr.ValueForProtocol(Dns4Protocol.Code)
-	if err != nil {
-		return nil, fmt.Errorf("error resolving %s: %s", maddr.String(), err)
-	}
-
-	encap := ma.Split(maddr)[1:]
-
-	result := []ma.Multiaddr{}
-	records, err := r.Backend.LookupIPAddr(ctx, value)
-	if err != nil {
-		return result, err
-	}
-
-	for _, r := range records {
-		ip4 := r.IP.To4()
-		if ip4 == nil {
-			continue
-		}
-		ip4maddr, err := ma.NewMultiaddr("/ip4/" + ip4.String())
-		if err != nil {
-			return result, err
+	var results []ma.Multiaddr
```

Comment:
Support resolving `/dns[46]` anywhere in an address, multiple times. Currently, we'll only resolve this at the beginning of the address.